### PR TITLE
Simplify ErtWorkflowList workflow structures

### LINKED ISSUE #4209: Simplify ErtWorkflowList

### PR DESCRIPTION
**Pull request by andreas-el**
**Issue**
Resolves #4209 

Still Todo:
- [x] Improve __repr__
- [x] Improve __eq__
- [x] Squash latest commits

PR in Everest to resolve blocking unit-test printout-issue:
https://github.com/equinor/everest/pull/1708

```
>       self.assertTrue(
            string_exists_in_file(everest_server_log_path, "Adding workflow job")
        )
E       AssertionError: False is not true
```

Simplify workflows, workflow_jobs, hook_workflows and workflow_job_dir
structures to use equally good but vastly more simple structures.
Extend tests to provide better coverage.
Refactor to improve syntax and readability.
Remove underlying obsoleted codeparts

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (oysteoh) [inline src/ert/_c_wrappers/enkf/ert_workflow_list.py:104]: does actually `addJob` return anything?

Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:104]: It should have, but lied about that! Crude fix in next commit. Good catch!

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (oysteoh) [inline src/ert/_c_wrappers/enkf/ert_workflow_list.py:184]: should this be `return job_name in self._workflow_jobs.keys()`  ?  I guess keys is a function? :thinking:

Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:184]: Sloppy work this. I've fixed the other one as well. Thanks!

File: src/ert/_c_wrappers/enkf/config_content_transmutor.py
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/config_content_transmutor.py:47]: In current trunk, there is a new version of `config_content_as_dict` in a separate file. I think that covers your use-case.

Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/config_content_transmutor.py]: Yep, it is a near identical implementation. Worked fine! :+1:

**Review by DanSava — COMMENTED** [review]
I think it is going in the right direction :+1

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:170]: Solved as part of https://github.com/equinor/ert/pull/4272

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:81]: Do you mean that this is not used publicly? We have loads of tests that utilize this?

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:159]: I'd like to remove as much as possible here, -- was hoping to remove the underlying object altogether.
Can i remove the subst_list from this call? The hook_workflows does not have any functionality left here, so this should also be removed.

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (oysteoh) [inline src/ert/_c_wrappers/enkf/ert_workflow_list.py:129]: just an annoying and unnecessary comment from my point of view - could it be removed?

Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:129]: Sure. These are leftovers from when restructuring. Consider it done.

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:73]: Yep. That makes sense. 👍

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:72]: `ErtWorkflowList` takes the following args in the constructor 
```python 
    def __init__(
        self,
        subst_list=None,
        config_content=None,
        config_dict=None,
        site_config_content=None,
    )
```
and what we do with `config_content` and `site_config_content` is 
```
content_dict = config_content_as_dict(config_content, site_config_content)
```
so my point is that  ErtWorkflowList could just get a config_dict as part of the constructor like:
 ``` python
 def __init__(
        self,
        subst_list=None,
        config_dict=None,
    )
```
and do `content_dict = config_content_as_dict(config_content, site_config_content)` in ResConfig before initializing ErtWorkflowList

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:150]: >Can i remove the subst_list from this call?
Should be removed once #4272 is merged and you rebase (there might be conflicts but hopefully not many)

Comment (DanSava) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:150]: >The hook_workflows does not have any functionality left here, so this should also be removed

Not 100% sure about that seems there is some functionality left related to the `hook_workflows`

**Comment by oysteoh** [comment]
Jenkins: Test this please!

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:72]: Note: you might have to change the inputs to `ResConfig(config_dict=...)` in the tests, but that is ok as its only the tests that use this path.

Comment (andreas-el) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py]: Im unsure if the provided commit satisfied this discussion. Now res_config passes an actual dict via the constructor instead, but there is still args for both content types (dicts).

**Comment by andreas-el** [comment]
Jenkins retest this please!

**Comment by oysteoh** [comment]
From my point of view this looks really good right now considering the code -cleanup and refactoring. I don't have much more to add at this point, but due to me not knowing all the technicalities  i would like i.e @eivindjahren  to have a look before approval to make sure the actual logic make sense :+1:

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/ert_workflow_list.py:179]: The difference between `_handle_config_content_dict` and `_handle_config_dict` is just the lookup of `ConfigKeys.Name`, `ConfigKeys.Path`, and `ConfigKeys.RUNMODE` are indices instead. Because there is no code using `_handle_config_dict` (except for tests). I suggest we change the expected input for that to be integer indices instead of strings so that we can have just one code path.

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/ert_workflow_list.py:16]: Instead of passing dicts to `__init__`, I would prefer it if we pass `workflow_job_info`, `workflow_job_dir_info`, `hook_workflow_info` and `workflow_info` to `__init__` instead and create a classmethod `from_dict` which gets the dictionary and creates the class via `__init__`. See `analysis_config.py` for an example of what I mean.

**Review by eivindjahren — CHANGES_REQUESTED** [review]

**Comment by eivindjahren** [comment]
I had some comments for further changes to `__init__`. If you would rather do those changes in a separate PR, we can merge this now.

File: src/ert/_c_wrappers/enkf/ert_workflow_list.py
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/ert_workflow_list.py:179]: After talking in person, we decided to start a new PR for this.

**Review by eivindjahren — APPROVED** [review]
:rocket: :chair:

**Pull request merged**
